A15-binary-logging.md

```diff
@@ -0,0 +1,375 @@
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_SERVER_TRAILER = 6;
+    // Signal that the RPC is cancelled.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_CANCEL = 7;
+  }
+
+  // Enumerates the entity that generates the log entry
+  enum Logger {
+    LOGGER_UNKNOWN = 0;
+    LOGGER_CLIENT = 1;
+    LOGGER_SERVER = 2;
+  }
+
+  // The timestamp of the binary log message
+  google.protobuf.Timestamp timestamp = 1;
+
+  // Uniquely identifies a call. The value must not be 0 in order to disambiguate
+  // from an unset value.
+  // Each call may have several log entries, they will all have the same call_id.
+  // Nothing is guaranteed about their value other than they are unique across
+  // different RPCs in the same gRPC process.
+  int64 call_id = 2;
+
+  // The entry sequence id for this call. The first GrpcLogEntry has a
+  // value of 1, to disambiguate from an unset value. The purpose of
+  // this field is to detect missing entries in environments where
+  // durability or ordering is not guaranteed.
+  uint32 sequence_id_within_call = 3;
+
+  EventType type = 4;
+  Logger logger = 5;  // One of the above Logger enum
+
+  // The logger uses one of the following fields to record the payload,
+  // according to the type of the log entry.
+  oneof payload {
+    ClientHeader client_header = 6;
+    ServerHeader server_header = 7;
+    // Used by EVENT_TYPE_CLIENT_MESSAGE, EVENT_TYPE_SERVER_MESSAGE
+    Message message = 8;
+    Trailer trailer = 9;
+  }
+
+  // true if payload does not represent the full message or metadata.
+  bool payload_truncated = 10;
+
+  // Peer address information, will only be recorded in
+  // EVENT_TYPE_SERVER_HEADER for client side or
+  // EVENT_TYPE_CLIENT_HEADER for server side.
+  // On client side, the remote peer is not yet known when the header
+  // is sent because load balancing has not yet happened.
+  // On server side, it is possible that the RPC is cancelled before
+  // it sends the header.
+  Peer peer = 11;
+};
+
+message ClientHeader {
+  // This contains only the metadata from the application.
+  Metadata metadata = 1;
+
+  string method_name = 2;
+
+  string authority = 3;
+
+  // the RPC timeout
+  google.protobuf.Duration timeout = 4;
+
+}
+
+message ServerHeader {
+  // This contains only the metadata from the application.
+  Metadata metadata = 3;
+}
+
+message Trailer {
+  // status_code and status_message:
+  // Only present for EVENT_TYPE_SERVER_TRAILER.
+  uint32 status_code = 1;
+
+  // An original status message before any transport specific
+  // encoding.
+  string status_message = 2;
+
+  // The value of the 'grpc-status-details-bin' metadata key. If
+  // present, this is always an encoded 'google.rpc.Status' message.
+  bytes status_details = 3;
+}
```

Comment:
Should this have a Metadata field?

Reply:
Fixed, this slipped in while copy pasting the sub messages around